### PR TITLE
Add complete support for Ergast API

### PR DESCRIPTION
**Pull request by theOehrly**
Hi everybody,

this pull request intends to implement full support for the Ergast API (https://ergast.com/mrd/).
Before actually introducing such a large new feature, I'd like to get a bit of feedback. Because now it is still easy to change the interface slightly or choose some different names for functions, properties and so on. I took the liberty to mention a few users that were/are fairly active with issues, discussions or PRs and everybody else is welcome to leave feedback as well. Hopefully there are a few people who can test this over the next two weeks or so.

Documentation can be found here: https://theoehrly.github.io/Fast-F1-Pre-Release-Documentation/ergast.html
I've uploaded a pre-release with a build of this PR: https://github.com/theOehrly/Fast-F1/releases/tag/alpha-3.0.0_ergast_a0 
To install, download the *.whl file and run ``pip install <filename>.whl``.

## Overview over the features:

- all endpoints supported
- all query filters supported
- can return data as Pandas DataFrames or raw
- can do automatic type casting to the best suited data type
- uses caching
- supports result limits and pagination
- enforces rate limits defined by Ergast
- cached requests do not count towards rate limits

(Mentions: @oscr @transorsmth @bruzie @hash55 @AND2797 @TracingInsights)

### CODE REVIEW

**Comment by oscr** [comment]
Awesome work! Happy to see this feature and just before the 2023 season starts so great timing. I will try to get a look at it during next week(end?).

**Review by oscr — COMMENTED** [review]
Impression:
- I was a little surprised by `standings = ergast.get_driver_standings()` not giving the current year and needed explicit year.
- Needed to remove `fetch_weekend` to get it to work.  
- I got a crash when invoking get_driver_standings which seems to be tied to the old cache. Once the old cache was cleared the code worked.
- I didn't fully understand what return type `Union[ErgastMultiResponse, ErgastRawResponse]` means. I eventually worked around it by `standings.content[0]` but `len(standings.content)
1` there is somethings else you can do with it?
- I wrote a minor update to the who can win example available here if you want to cherry pick it 
https://github.com/theOehrly/Fast-F1/compare/ergast-proposal-1...oscr:Fast-F1:ergast-proposal-1

Overall I think this is a very promising addition. Great work!

File: fastf1/ergast/__init__.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import fastf1.ergast.interface
+
+# imports for exposed names
+from fastf1.ergast.interface import Ergast  # noqa: F401
+from fastf1.ergast.legacy import \
+    fetch_day, \
+    fetch_weekend, \
```
[the remark this answers is not in the document]
Comment (theOehrly) [reply fastf1/ergast/__init__.py:7]: I assume you tested this on the current PR branch and not with the wheel that I provided? I think I just messed this up with the last commit that I added after this. I'll need to merge master and then it should be fine.

**Comment by theOehrly** [comment]
>    I was a little surprised by `standings = ergast.get_driver_standings()` not giving the current year and needed explicit year.

I understand, but this is basically for API consitency reasons. Many API endpoints take ``season`` as an argument. And I don't think it should default to ``'current'`` for all of these. But then, if it only has a default for some endpoints, that would probably be just as unexpected.

>    Needed to remove `fetch_weekend` to get it to work.

Yes, messed this up by splitting commits between master and this branch, sorry.

>    I got a crash when invoking get_driver_standings which seems to be tied to the old cache. Once the old cache was cleared the code worked.

That is very weird. It never happend to me during developement. I hope that this would not be a widespread problem. I need to try and find an old cache somewhere on my machine and try and reproduce it.

>    I didn't fully understand what return type `Union[ErgastMultiResponse, ErgastRawResponse]` means. I eventually worked around it by `standings.content[0]` but `len(standings.content) 1` there is somethings else you can do with it?

Union just means that the return type is one of the types listed. This depends on whether you select ``return_type='pandas'`` or ``return_type='raw'``.
You get an ErgastMultiResponse... for consistency reasons because the structure is equivalent to other endpoints that return more data.
But there is also at least one possible combination of selections that returns multiple results. 

```
from fastf1.ergast import Ergast

ergast = Ergast()
ergast.get_driver_standings(driver='alonso').description
#     season  round
# 0     2001     17
# 1     2003     16
# 2     2004     18
# 3     2005     19
# ...
len(ergast.get_driver_standings(driver='alonso').content)
# 19
```
Now, for this response, each DataFrame in ``.content`` only has one row, which is somewhat unfortunate. But the alternative is to introduce yet another response type and then it is even less intuitive what kind of data structure you'll get back when you use the pandas result type. At least I don't have a better idea. 
So again, for consistency, I went with this approach. If you have an idea how to do this better, let me know.
 
>    I wrote a minor update to the who can win example available here if you want to cherry pick it
>    [ergast-proposal-1...oscr:Fast-F1:ergast-proposal-1](https://github.com/theOehrly/Fast-F1/compare/ergast-proposal-1...oscr:Fast-F1:ergast-proposal-1)

I'll take a look at that as soon as I have time for that. Thanks a lot!

**Comment by oscr** [comment]
I would like to clarify when I say "I was surprised" that doesn't mean there is anything wrong with this pr. It's should be interpreted as "given my limited knowledge of Pandas etc I expected it to work in a certain way. Which it may not because I don't fully understand the domain".

> > I was a little surprised by `standings = ergast.get_driver_standings()` not giving the current year and needed explicit year.
> 
> I understand, but this is basically for API consitency reasons. Many API endpoints take `season` as an argument. And I don't think it should default to `'current'` for all of these. But then, if it only has a default for some endpoints, that would probably be just as unexpected.

Maybe this is something we can highlight in examples by having clear SEASON variables and/or adding it to the docs: https://theoehrly.github.io/Fast-F1-Pre-Release-Documentation/ergast.html#fastf1.ergast.Ergast.get_driver_standings
session default value: 19XX

> > Needed to remove `fetch_weekend` to get it to work.
> 
> Yes, messed this up by splitting commits between master and this branch, sorry.

No need to apologize. Release candidates are expected to have bugs :)

> > I got a crash when invoking get_driver_standings which seems to be tied to the old cache. Once the old cache was cleared the code worked.
> 
> That is very weird. It never happend to me during developement. I hope that this would not be a widespread problem. I need to try and find an old cache somewhere on my machine and try and reproduce it.

I deleted it by hand but I'm assuming `Cache.clear_cache(deep=True)` would also solve it. So might be worth keeping in mind or having as a troubleshooting step for upgrading. But it could have just been something with my machine.

> > I didn't fully understand what return type `Union[ErgastMultiResponse, ErgastRawResponse]` means. I eventually worked around it by `standings.content[0]` but `len(standings.content) 1` there is somethings else you can do with it?
> 
> Union just means that the return type is one of the types listed. This depends on whether you select `return_type='pandas'` or `return_type='raw'`. You get an ErgastMultiResponse... for consistency reasons because the structure is equivalent to other endpoints that return more data. But there is also at least one possible combination of selections that returns multiple results.

Thank you for explaining this. Could this maybe be highlighted in the docs? 

> ```
> from fastf1.ergast import Ergast
> 
> ergast = Ergast()
> ergast.get_driver_standings(driver='alonso').description
> #     season  round
> # 0     2001     17
> # 1     2003     16
> # 2     2004     18
> # 3     2005     19
> # ...
> len(ergast.get_driver_standings(driver='alonso').content)
> # 19
> ```
> 
> Now, for this response, each DataFrame in `.content` only has one row, which is somewhat unfortunate. But the alternative is to introduce yet another response type and then it is even less intuitive what kind of data structure you'll get back when you use the pandas result type. At least I don't have a better idea. So again, for consistency, I went with this approach. If you have an idea how to do this better, let me know.
> 
> > I wrote a minor update to the who can win example available here if you want to cherry pick it
> > [ergast-proposal-1...oscr:Fast-F1:ergast-proposal-1](https://github.com/theOehrly/Fast-F1/compare/ergast-proposal-1...oscr:Fast-F1:ergast-proposal-1)
> 
> I'll take a look at that as soon as I have time for that. Thanks a lot!

Thank you for the great feedback! Great work with this

**Comment by theOehrly** [comment]
Thank you @oscr for this valuable feedback! I will go through the documentation again and try to clarify things a bit more. Especially the points that you noted.

**Comment by theOehrly** [comment]
I reworked the docs a bit more finally. And I cherry-picked your change to the example code @oscr. Thank you very much for that and for your feedback!

**Pull request merged**